protocol-units/bridge/chains/ethereum/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+	>,
+	RootProvider<BoxTransport>,
+	BoxTransport,
+	Ethereum,
+>;
+
+#[derive(RlpDecodable, RlpEncodable)]
+struct EthBridgeTransferDetails {
+	pub amount: U256,
+	pub originator: EthAddress,
+	pub recipient: [u8; 32],
+	pub hash_lock: [u8; 32],
+	pub time_lock: U256,
+	pub state: u8, // Assuming the enum is u8 for now..
+}
+
+pub struct EthClient<P> {
+	rpc_provider: P,
+	chain_id: String,
+	ws_provider: RootProvider<PubSubFrontend>,
+	initiator_address: EthAddress,
+	counterparty_address: EthAddress,
+	send_transaction_error_rules: Vec<Box<dyn VerifyRule>>,
+	gas_limit: u64,
+	num_tx_send_retries: u32,
+}
+
+impl EthClient<AlloyProvider> {
+	pub async fn build_with_config(
+		config: Config,
+		counterparty_address: &str,
+	) -> Result<Self, anyhow::Error> {
+		let signer_private_key = config.signer_private_key;
+		let signer = signer_private_key.parse::<PrivateKeySigner>()?;
+		let initiator_address = config.initiator_address.parse()?;
+		let rpc_url = config.rpc_url.context("rpc_url not set")?;
+		let ws_url = config.ws_url.context("ws_url not set")?;
+		let rpc_provider = ProviderBuilder::new()
+			.with_recommended_fillers()
+			.wallet(EthereumWallet::from(signer.clone()))
+			.on_builtin(&rpc_url)
+			.await?;
+		let ws = WsConnect::new(ws_url);
+		let ws_provider = ProviderBuilder::new().on_ws(ws).await?;
+
+		EthClient::build_with_provider(utils::ProviderArgs {
+			rpc_provider,
+			ws_provider,
+			initiator_address,
+			counterparty_address: counterparty_address.parse()?,
+			gas_limit: config.gas_limit,
+			num_tx_send_retries: config.num_tx_send_retries,
+			chain_id: config.chain_id,
+		})
+		.await
+	}
+
+	async fn build_with_provider(args: utils::ProviderArgs) -> Result<Self, anyhow::Error> {
+		let rule1: Box<dyn VerifyRule> = Box::new(SendTransactionErrorRule::<UnderPriced>::new());
+		let rule2: Box<dyn VerifyRule> =
+			Box::new(SendTransactionErrorRule::<InsufficentFunds>::new());
+		let send_transaction_error_rules = vec![rule1, rule2];
+		Ok(EthClient {
+			rpc_provider: args.rpc_provider,
+			chain_id: args.chain_id,
+			ws_provider: args.ws_provider,
+			initiator_address: args.initiator_address,
+			counterparty_address: args.counterparty_address,
+			gas_limit: args.gas_limit,
+			num_tx_send_retries: args.num_tx_send_retries,
+			send_transaction_error_rules,
+		})
+	}
+}
+
+impl<P> Clone for EthClient<P> {
+	fn clone(&self) -> Self {
+		todo!()
+	}
+}
+
+#[async_trait::async_trait]
+impl<P> BridgeContractInitiator for EthClient<P>
+where
+	P: Provider + Clone + Send + Sync + Unpin,
+{
+	type Address = EthAddress;
+	type Hash = EthHash;
+
+	async fn initiate_bridge_transfer(
+		&mut self,
+		_initiator_address: InitiatorAddress<Self::Address>,
+		recipient_address: RecipientAddress<Vec<u8>>,
+		hash_lock: HashLock<Self::Hash>,
+		time_lock: TimeLock,
+		amount: Amount,
+	) -> BridgeContractInitiatorResult<()> {
+		let contract = AtomicBridgeInitiator::new(self.initiator_address.0, &self.rpc_provider);
```

Comment:
Initiator address from the point of the trait impls is the signer, not the initiator contract address